perf/perf.go

```diff
@@ -0,0 +1,520 @@
+// number of lost samples.
+type Record struct {
+	// The data submitted via bpf_perf_event_output.
+	// They are padded with 0 to have a 64-bit alignment.
+	// If you are using variable length samples you need to take
+	// this into account.
+	RawSample []byte
+
+	// The number of samples which could not be output, since
+	// the ring buffer was full.
+	LostSamples uint64
+}
+
+func readRecord(rd io.Reader) (*Record, error) {
+	const (
+		perfRecordLost   = 2
+		perfRecordSample = 9
+	)
+
+	var header perfEventHeader
+	err := binary.Read(rd, internal.NativeEndian, &header)
+	if err == io.EOF {
+		return nil, nil
+	}
+
+	if err != nil {
+		return nil, errors.Wrap(err, "can't read event header")
+	}
+
+	switch header.Type {
+	case perfRecordLost:
+		lost, err := readLostRecords(rd)
+		if err != nil {
+			return nil, err
+		}
+
+		return &Record{LostSamples: lost}, nil
+
+	case perfRecordSample:
+		raw, err := readRawSample(rd)
+		if err != nil {
+			return nil, err
+		}
+
+		return &Record{RawSample: raw}, nil
+
+	default:
+		return nil, errors.Errorf("unknown event type %d", header.Type)
+	}
+}
+
+func readLostRecords(rd io.Reader) (uint64, error) {
+	var lostHeader struct {
+		ID   uint64
+		Lost uint64
+	}
+
+	err := binary.Read(rd, internal.NativeEndian, &lostHeader)
+	if err != nil {
+		return 0, errors.Wrap(err, "can't read lost records header")
+	}
+
+	return lostHeader.Lost, nil
+}
+
+func readRawSample(rd io.Reader) ([]byte, error) {
+	var size uint32
+	if err := binary.Read(rd, internal.NativeEndian, &size); err != nil {
+		return nil, errors.Wrap(err, "can't read sample size")
+	}
+
+	data := make([]byte, int(size))
+	_, err := io.ReadFull(rd, data)
+	return data, errors.Wrap(err, "can't read sample")
+}
+
+// Reader allows reading bpf_perf_event_output
+// from user space.
+type Reader struct {
+	mu sync.Mutex
+
+	// Closing a PERF_EVENT_ARRAY removes all event fds
+	// stored in it, so we keep a reference alive.
+	array *ebpf.Map
+	rings []*perfEventRing
+
+	epollFd     int
+	epollEvents []unix.EpollEvent
+	epollRings  []*perfEventRing
+	// Eventfds for closing
+	closeFd int
+	// Ensure we only close once
+	closeOnce sync.Once
+}
+
+// ReaderOptions control the behaviour of the user
+// space reader.
+type ReaderOptions struct {
+	// A map of type PerfEventArray.
+	Map *ebpf.Map
+	// Controls the size of the per CPU buffer in bytes.
+	PerCPUBuffer int
+	// The number of written bytes required in any per CPU buffer before
+	// Read will process data. Must be smaller than PerCPUBuffer.
+	Watermark int
+}
+
+const (
+	// Sentinel value passed in EpollEvent.Pad
+	wokenByClose = 1
+)
+
+// NewReader creates a new reader with the given options.
+func NewReader(opts ReaderOptions) (pr *Reader, err error) {
+	if opts.PerCPUBuffer < 1 {
+		return nil, errors.New("PerCPUBuffer must be larger than 0")
+	}
+
+	// We can't create a ring for CPUs that aren't online, so use only the online (of possible) CPUs
+	nCPU, err := internal.OnlineCPUs()
```

Comment:
I just read the documentation for `NumCPU`, and it seems to me like that value can be influenced by things like CPU affinity for the userspace proess. We really want online CPUs in the system, since we don't control what CPU an eBPF program runs on.